iotest/iotest.go

```diff
@@ -0,0 +1,92 @@
+// Package iotest implements io Read/Writers that are useful for testing.
+// It is inspired on the stdlib iotest: https://golang.org/pkg/testing/iotest/
+// It adds on it, instead of being a replacement.
+package iotest
+
+import (
+	"fmt"
+	"io"
+)
+
+// RepeatReader is an io.Reader that repeats a given io.Reader
+// It is NOT safe to use a RepeatReader concurrently.
+type RepeatReader struct {
+	reader      io.Reader
+	readData    []byte
+	readIndex   int
+	err         error
+	repeatCount int
+}
+
+// RepeatReaderError is an enumeration of errors that are part of
+// RepeatReader interface. The sentinel errors are wrapped
+// with dynamic context, so always check using errors.Is.
+type RepeatReaderError string
+
+const (
+	RepeatReaderInvalidCountErr RepeatReaderError = "Repeat reader count must be >= 0"
+)
+
+// NewRepeatReader creates RepeatReader that will repeat the
+// given io.Reader "n" times.
+//
+// If n=0 it won't provide any data and will return an io.EOF immediately (empty stream).
+// If n=1 it will repeat once, just like reading the original stream.
+// If n=2 it will repeat the underlying stream twice, doubling it.
+//
+// It will use O(N) memory where N is the size of the contents read from
+// the given reader (all contents are kept in memory and looped over).
+//
+// It can be a cheap way to generate gigantic inputs by repeating a very
+// small input (it will only use O(N) where N= small input size).
+//
+// It is a severe programming error to pass a negative count, resulting in a panic.
+func NewRepeatReader(r io.Reader, n int) *RepeatReader {
+	if n < 0 {
+		panic(fmt.Errorf("%w : count is %d", RepeatReaderInvalidCountErr, n))
+	}
+	return &RepeatReader{
+		reader:      r,
+		repeatCount: n,
+	}
+}
+
+// Read implements the same behavior of the underlying io.Reader + repeating
+// semantics.
+func (r *RepeatReader) Read(d []byte) (int, error) {
+	if r.repeatCount == 0 {
+		if r.err == nil {
+			r.err = io.EOF
```

Comment:
since it should always return `io.EOF`, why check the error for nil?

Reply:
hmm I think I had a mental model that I should respect whatever was stored on r.err when returning results. But indeed if we reach repeatCount == 0 it is always because it is an EOF, when we have errors the counter is not decremented anyway (so we never reach 0), and when it is initialized with 0 then we want EOF in that case too...so indeed I think we can simplify this. Thanks for the catch :-)